storage: check leader not leaseholder condition on all commands

spencerkimball opened the pull request:
Previously, we only checked for leader-not-leaseholder when completing a
lease operation. This relied on the constant renewals of expiration-based
leases to transfer Raft leadership reliably.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12323)
<!-- Reviewable:end -->

spencerkimball committed:
storage: correct leader-not-leaseholder condition

Previously, we only checked for leader-not-leaseholder when completing a
lease operation. This relied on the constant renewals of expiration-based
leases to transfer Raft leadership reliably. We now do this while checking
if a range can be quiesced. This is convenient because we want to also
avoid quiescing if we notice this problem.

Also, only attempt to transfer raft leader to the leaseholder if the
leaseholder (the transferee) has fully applied the raft log. Otherwise,
the leader will send the presumably unapplied log suffix to the transferee
again and again.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtW6Au6q2SSLbbJxQ7:-KYtW6Au6q2SSLbbJxQ8:b3hat2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2917)):*
> ```Go
> 
> 	status := repl0.RaftStatus()
> 	if status != nil && status.Lead != 1 {
> ```

while you're here `nil` should probably not be allowed 

---

*[pkg/storage/client_raft_test.go, line 2921 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtVjHuUgXEQ5ckpzwe:-KYtVjHuUgXEQ5ckpzwf:bhxu5ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2921)):*
> ```Go
> 	}
> 
> 	// Force a read on Store 2 to request a new lease. Other moving parts in
> ```

this is no longer testing the mechanism by which raft leadership transfers happen, right? seems odd to see this code here now that transfers are triggered by all commands.

on second thought, maybe it is just that this comment is outdated.

---

*[pkg/storage/client_raft_test.go, line 2943 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUzl_8qh12TdOQr34:-KYtUzl_8qh12TdOQr35:bxrpg6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2943)):*
> ```Go
> 	// Wait for raft leadership transferring to be finished.
> 	testutils.SucceedsSoon(t, func() error {
> 		status = repl0.RaftStatus()
> ```

nit: `:=`

---

*[pkg/storage/client_raft_test.go, line 2963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtVw_oGsqojv-ayKjC:-KYtVw_oGsqojv-ayKjD:bxrpg6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2963)):*
> ```Go
> 			return err
> 		}
> 		status = repl1.RaftStatus()
> ```

nit: `:=`

---

*[pkg/storage/client_raft_test.go, line 2964 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtWY_HGpmtT_Dt6wLy:-KYtWY_HGpmtT_Dt6wLz:b-w1wvoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2964)):*
> ```Go
> 		}
> 		status = repl1.RaftStatus()
> 		if status.Lead != 1 {
> ```

nit: this `1` should probably be `rd0.ReplicaID`?

---

*[pkg/storage/client_raft_test.go, line 2982 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtW2YhIbC7OTSRRQKo:-KYtW2YhIbC7OTSRRQKp:bxrpg6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2982)):*
> ```Go
> 			return pErr.GetDetail()
> 		}
> 		status = repl0.RaftStatus()
> ```

nit: `:=`

---

*[pkg/storage/helpers_test.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUMvIC9YG3jaw0kwj:-KYtUMvJACX33AnMSIhJ:bb7kgm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/helpers_test.go#L264)):*
> ```Go
> // command will silently fail if the target is not an up-to-date
> // replica.
> func (r *Replica) RaftTransferLeader(target roachpb.ReplicaID) error {
> ```

this function is very similar to `(*Replica).maybeTransferRaftLeadership` - could you reduce the probability of rot by combining (or extracting a common method used by both)?

---

*[pkg/storage/replica.go, line 3473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtU2hhRn5Ppg7JO3G_:-KYtU2hhRn5Ppg7JO3Ga:b-g1h1gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica.go#L3473)):*
> ```Go
> 		// quick command application (requests generally need to make it to both the
> 		// lease holder and the raft leader before being applied by other replicas).
> 		if raftCmd.BatchRequest != nil {
> ```

why the nil check?

---

*[pkg/storage/replica_proposal.go, line 430 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUjOtIZPOVSjH5Jov:-KYtUjOtIZPOVSjH5Jow:b-9604zz) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L430)):*
> ```Go
> 			if l := r.mu.state.Lease; !l.OwnedBy(r.store.StoreID()) &&
> 				r.isLeaseValidLocked(l, r.store.Clock().Now()) {
> 				log.Infof(ctx, "transferring raft leadership to replica ID %v", l.Replica.ReplicaID)
> ```

while you're here, log the whole target replica? `log.Infof(ctx, "transferring raft leadership to %+v", l.Replica)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYtWjo6LM5TBJz7lI-K:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
In that past when this came up there was concern about attempting too many of these raft transfers in case they're constantly failing. I don't understand how these transfers work and under which conditions they might fail, but maybe some throttling is in order?

Btw, @ben, @petermattis, is it time to give https://reviewable.io/reviews/cockroachdb/cockroach/8837 another look? It was trying to address #8816 by transferring raft leadership before transferring leases.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYtl_Y_HPYibwJDPaOM:bk58gpf)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtW6Au6q2SSLbbJxQ7:-KYtcr6ddxJxGqTGLEuU:baosfxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2917)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

while you're here `nil` should probably not be allowed 
</blockquote></details>

Right: `status == nil || status.Lead != 1`

---

*[pkg/storage/replica_proposal.go, line 429 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtdSTmytgYef5kj_Sh:-KYtdSTmytgYef5kj_Si:b-5hrwee) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L429)):*
> ```Go
> 		if raftGroup.Status().RaftState == raft.StateLeader {
> 			if l := r.mu.state.Lease; !l.OwnedBy(r.store.StoreID()) &&
> 				r.isLeaseValidLocked(l, r.store.Clock().Now()) {
> ```

Can we do the lease check first? `raftGroup.Status()` feels a bit more expensive and the order of the checks shouldn't matter. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

The concern about leadership transfers failing was partially FUD, and I'm less concerned about it now since this appears to have been working reliably so far. Transfers can fail if network messages get reordered, but that shouldn't happen with our raft transport. We should have a metric for how often these transfers are attempted, but I don't think we need to build throttling unless we see a problem. (if anything, throttling could mask a problem by allowing a range to limp along with limited availability; without throttling the problem will be more visible). 

The *combination* of this change and #8837 would be bad: the node that receives the leadership transfer under #8837 would be likely to transfer it right back to the former leader under this change (and I think that's more of an argument against #8837 than against this change). I'll take a look at #8837 and #8816. 

The commit message should be updated to say "on all write commands"; read-only traffic won't trigger this check. Should it? It doesn't matter in practice since a stable range with read-only traffic doesn't really care whether the lease holder and raft leader are the same, but it does mean that the leader-not-leaseholder metric wouldn't reliably fall to zero. (for that matter, the range could be completely idle. Should we have a scanner queue to perform this transfer as a time-based last resort?)

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 431 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYvJklOwvAs_3JSELA4:-KYvJklPixn8Ko_zbOmz:b-ekppq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L431)):*
> ```Go
> 				r.isLeaseValidLocked(l, r.store.Clock().Now()) {
> 				log.Infof(ctx, "transferring raft leadership to replica ID %v", l.Replica.ReplicaID)
> 				raftGroup.TransferLeader(uint64(l.Replica.ReplicaID))
> ```

We should increment a metric here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYvGaCr6hMQ-DxPfcl6:b-g4rk6f)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
I've changed this to instead tee these up during metrics computation when we realize the condition exists. Otherwise, I keep seeing small numbers of long-running replicas with the leader-not-leaseholder condition.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtW6Au6q2SSLbbJxQ7:-KYu_Tj5vVTOUHSxKnpX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2917)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Right: `status == nil || status.Lead != 1`
</blockquote></details>

Done.

---

*[pkg/storage/client_raft_test.go, line 2921 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtVjHuUgXEQ5ckpzwe:-KYxt2dqUDdJ5C5U6dAn:b-ev2pdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2921)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is no longer testing the mechanism by which raft leadership transfers happen, right? seems odd to see this code here now that transfers are triggered by all commands.

on second thought, maybe it is just that this comment is outdated.
</blockquote></details>

I've mostly restored this unittest as I decided to make the transfers happen during metrics computation instead.

---

*[pkg/storage/client_raft_test.go, line 2943 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUzl_8qh12TdOQr34:-KYxtAT3aAVUt-dabSlp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2943)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `:=`
</blockquote></details>

Done.

---

*[pkg/storage/client_raft_test.go, line 2963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtVw_oGsqojv-ayKjC:-KYxtCuAvs0FOoySqCup:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2963)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `:=`
</blockquote></details>

Done.

---

*[pkg/storage/client_raft_test.go, line 2964 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtWY_HGpmtT_Dt6wLy:-KYxtMd4ZjwzoClhLxsn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2964)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: this `1` should probably be `rd0.ReplicaID`?
</blockquote></details>

Done.

---

*[pkg/storage/client_raft_test.go, line 2982 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtW2YhIbC7OTSRRQKo:-KYxtN2G9HYupJ38j_ww:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/client_raft_test.go#L2982)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `:=`
</blockquote></details>

Done.

---

*[pkg/storage/helpers_test.go, line 264 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUMvIC9YG3jaw0kwj:-KYxtmZTSr5yd8ADbdvs:b-tz8jyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/helpers_test.go#L264)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this function is very similar to `(*Replica).maybeTransferRaftLeadership` - could you reduce the probability of rot by combining (or extracting a common method used by both)?
</blockquote></details>

It's simplified now.

---

*[pkg/storage/replica.go, line 3473 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtU2hhRn5Ppg7JO3G_:-KYxto4QUvfkI3OgekMc:b-87xzux) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica.go#L3473)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why the nil check?
</blockquote></details>

This is gone now. I was avoiding doing making this check after Raft leadership transfers, which always send a nil request. It made testing impossible.

---

*[pkg/storage/replica_proposal.go, line 429 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtdSTmytgYef5kj_Sh:-KYxtuh__Ua3St3UxWtv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L429)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can we do the lease check first? `raftGroup.Status()` feels a bit more expensive and the order of the checks shouldn't matter. 
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 430 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYtUjOtIZPOVSjH5Jov:-KYxtwPjzHvP0eHyqw0d:btsfsoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L430)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

while you're here, log the whole target replica? `log.Infof(ctx, "transferring raft leadership to %+v", l.Replica)`
</blockquote></details>

I've reverted to old code here.

---

*[pkg/storage/replica_proposal.go, line 431 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYvJklOwvAs_3JSELA4:-KYxu-OJvY40l_8dpUjE:byyyvhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L431)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We should increment a metric here. 
</blockquote></details>

We already have a `raft.rcvd.transferleader` which is incremented elsewhere. Do you think we need both? In any case, I've reverted the changes to this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYxuUjGetdg2WhqT3RD:bcz80zq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 4472 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYy3AlEHuRieBNBSG8-:-KYy3AlEHuRieBNBSG80:b-b0v3jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/854703f3be63f48a1c4ff7f53fbd574a17a95437/pkg/storage/replica.go#L4472)):*
> ```Go
> 	// Try to correct leader-not-leaseholder condition, if encountered.
> 	if m.leaseValid && !m.leaseholder && m.leader {
> 		r.maybeTransferRaftLeadership(ctx, status.lease.Replica.ReplicaID)
> ```

It's very strange for a method named `metrics` to have the side effect of transferring the Raft leadership. Can we pull this up to the caller? Even doing this in the caller is a bit strange, though not quite so egregious. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2966 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYy3xzS0XfdtWVHJ-eB:-KYy3xzS0XfdtWVHJ-eC:b8otl5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/854703f3be63f48a1c4ff7f53fbd574a17a95437/pkg/storage/client_raft_test.go#L2966)):*
> ```Go
> 		status := repl1.RaftStatus()
> 		if status.Lead != rd0.ReplicaID {
> 			return errors.Errorf("expected raft leader be 1; got %d", status.Lead)
> ```

this "1" is still hard coded

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2966 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYy3xzS0XfdtWVHJ-eB:-KYyVczO6yWjWhSG-JR5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/854703f3be63f48a1c4ff7f53fbd574a17a95437/pkg/storage/client_raft_test.go#L2966)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this "1" is still hard coded
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 4472 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYy3AlEHuRieBNBSG8-:-KYyVd_ZGGzy_6r5W3EA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/854703f3be63f48a1c4ff7f53fbd574a17a95437/pkg/storage/replica.go#L4472)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It's very strange for a method named `metrics` to have the side effect of transferring the Raft leadership. Can we pull this up to the caller? Even doing this in the caller is a bit strange, though not quite so egregious. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
This PR causes the cluster to lock up. Need to investigate why, but my suspicion is there's something happening in `raft.RawNode.TransferLeader`.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYyYQCvahoJzo29-6zH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 2944 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYyuDZU4k-83o6npFNF:-KYyuDZVYh76aR9hlrLd:b-vfmbob) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e4e348c58c8dbb8c5e38afe8ee172f55a9f87ee/pkg/storage/client_raft_test.go#L2944)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		status := repl0.RaftStatus()
> 		if status.Lead != 2 {
> ```

bunch more hardcoded IDs here, but not a blocker.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KYyuHiW5xSp6XNjbpFR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2944 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYyuDZU4k-83o6npFNF:-KYz3yE1rNTVRPnM-76X:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e4e348c58c8dbb8c5e38afe8ee172f55a9f87ee/pkg/storage/client_raft_test.go#L2944)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

bunch more hardcoded IDs here, but not a blocker.
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KYzDDsOESuWx22lgSbE:bkzsqs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
> ```Go
> 
> 	status := repl0.RaftStatus()
> 	if status == nil || status.Lead != 1 {
> ```

this "1" (and on the next line) is rd0.ReplicaID. should be easy now that you've moved the definition of `rd0` to the top

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDRYYiM64ez3B9eal:-KYzDRYZBGigcoVWwygc:bl8z8o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
> ```Go
> 		// Only the raft leader can attempt a leadership transfer.
> 		if status := raftGroup.Status(); status.RaftState == raft.StateLeader {
> 			// Only attempt this if the leaseholder has all the log entries.
> ```

did this solve the cluster lockup problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 440 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDq-ktC_KyCP520r9:-KYzDq-ktC_KyCP520rA:b-728p8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L440)):*
> ```Go
> 		if status := raftGroup.Status(); status.RaftState == raft.StateLeader {
> 			// Only attempt this if the leaseholder has all the log entries.
> 			if pr, ok := status.Progress[uint64(target)]; ok && pr.Match == r.mu.lastIndex {
> ```

Unless I'm misreading, the check in the raft code is `status.Progress[leader] == r.mu.lastIndex`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzEDbW_rkKgXb_3Z7D:-KYzEDbW_rkKgXb_3Z7E:b6ryhg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
> ```Go
> 		// Only the raft leader can attempt a leadership transfer.
> 		if status := raftGroup.Status(); status.RaftState == raft.StateLeader {
> 			// Only attempt this if the leaseholder has all the log entries.
> ```

s/leaseholder/target/ - only the caller knows that the target is the leaseholder

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KYzZAeUIztIg8c7g-IT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this "1" (and on the next line) is rd0.ReplicaID. should be easy now that you've moved the definition of `rd0` to the top
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDRYYiM64ez3B9eal:-KYzZBO-3Ri9y8ZcNk1a:b-kwgq2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did this solve the cluster lockup problem?
</blockquote></details>

No. Still looking. This PR will continue to change.

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzEDbW_rkKgXb_3Z7D:-KYzZRBG20nTXGXbOg1N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/leaseholder/target/ - only the caller knows that the target is the leaseholder
</blockquote></details>

Done.

---

*[pkg/storage/replica_proposal.go, line 440 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDq-ktC_KyCP520r9:-KYzZFWRNqdRxl-xXY-H:bm2cxgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L440)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Unless I'm misreading, the check in the raft code is `status.Progress[leader] == r.mu.lastIndex`. 
</blockquote></details>

You're misreading. The `msg.From` field is set to the transferee, and that is the progress which is checked in `raft.stepLeader`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
My guess is that quiescence is preventing a leadership transfer from being canceled in https://github.com/coreos/etcd/blob/d8513adf1d2058047929b1ed7c038fbe023ebe3a/raft/raft.go#L548 (but heartbeats are still flowing and preventing anyone but our designated target from campaigning). This could occur whenever we transfer leadership to a node that is going away, but this was unlikely when we only transferred to a node that successfully proposed and acquired a lease. Here, it's possible that we're transferring leadership to a node just as its replica is being removed. I think we need to prevent quiescence when a leader transfer is in progress.

---

Reviewed 2 of 4 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 431 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYvJklOwvAs_3JSELA4:-KZ-fJh5zwgjPX9_d2-b:b-grssdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L431)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

We already have a `raft.rcvd.transferleader` which is incremented elsewhere. Do you think we need both? In any case, I've reverted the changes to this file.
</blockquote></details>

`raft.rcvd.transferleader` is always zero because `MsgTransferLeader` is used internally by raft and not sent on the wire. `raft.rcvd.timeoutnow` should work, though (but it's not quite the same, since some calls to TransferLeader may fail without sending MsgTimeoutNow).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KZ-jwNgMTTSwm7aIGEb:b-34vt1o)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@bdarnell I'm not following how your scenario could occur. The leaseholder is the one which performs replica rebalancing for a range and it will never remove itself from the range. Instead, it first transfers its lease somewhere else and the recipient will perform the removal. I suppose it is possible for this lease transfer and removal to happen fast enough the the raft leadership transfer overlaps, but that seems unlikely. That said, I'll play around with some scenarios here as clearly we're seeing something on `blue` which seems to disappear when quiescence is disabled.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KZ1wgoItCF_PPzgphKf:b-ameop7)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Maybe it's racing with an addition instead of a removal, then: We see that the local node is overloaded while another node is underloaded. We rebalance a replica onto it, then as soon as the ChangeReplicas commits and we've processed it locally (but the target node hasn't), we transfer the lease, and then the raft leadership, all before that node has applied the EndTransaction that adds it (so it drops the raft MsgTimeoutNow message). The timing on that is really tight, but it could be possible, and that would have the same quiescence problem. There might be some quirk of preemptive snapshot processing (and its aftermath) that makes the timing more likely.

spencerkimball commented:
Damn the problem disappeared after the run where I changed `redirectOnOrAcquireLease` to always load the Raft group. In other words, that change fixed the problem permanently. I'd also experimented with always incrementing the liveness epoch and that presumably also should have fixed it but it didn't seem to; perhaps I just did something wrong. Unfortunately it's too late to go back and debug further. It seems as if some kind of shutdown scenario left one replica believing an old lease was active and then one replica and the other two played ping pong with lease expectation.

spencerkimball commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_proposal.go, line 431 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYvJklOwvAs_3JSELA4:-KZDVuFptYo58qEJY3Pe:by5lu7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/55df017e8aad8c24b8c797eeefdf6fa57b51e2f7/pkg/storage/replica_proposal.go#L431)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`raft.rcvd.transferleader` is always zero because `MsgTransferLeader` is used internally by raft and not sent on the wire. `raft.rcvd.timeoutnow` should work, though (but it's not quite the same, since some calls to TransferLeader may fail without sending MsgTimeoutNow).
</blockquote></details>

OK, added a metric.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KZD_5yKvpaEqu-h73yc:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

?

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzEDbW_rkKgXb_3Z7D:-KZD_CIF7P62SdxU2Ohr:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KZDofPHqOpqvjUW6cyV:b-wsihrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

?
</blockquote></details>

Damn, my changes here got reverted accidentally.

---

*[pkg/storage/replica_proposal.go, line 439 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzEDbW_rkKgXb_3Z7D:-KZDp7vd60oVVygiMpgb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/replica_proposal.go#L439)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KZDpiX-IoByjmfAYnae:b-o787pz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Damn, my changes here got reverted accidentally.
</blockquote></details>

the error message below still says "1", but lg otherwise

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
@petermattis take a look at the most recent change. I've decided to check for the condition at `maybeQuiesce` time because I realized during my testing this weekend that we should not quiesce ranges where the leaseholder and leader are the same replica.

spencerkimball commented:
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 2930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-KYzDDsOESuWx22lgSbD:-KZOGS55h2E2WW7QQS83:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9135693b0f9dca53de8f5e2626326e59e33208/pkg/storage/client_raft_test.go#L2930)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

the error message below still says "1", but lg otherwise
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r5, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12323#-:-KZQ4CX7zMj_XTC8FnlS:bnfp4nl)*
<!-- Sent from Reviewable.io -->
